hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/schema/TypeUtils.java

```diff
@@ -0,0 +1,328 @@
+    }
+
+    private static class PortableEnricher extends FieldEnricher<PortableId, ClassDefinition> {
+        private final PortableContext context;
+
+        PortableEnricher(RelationsStorage relationsStorage, InternalSerializationService serializationService) {
+            super(QueryDataType.OBJECT_TYPE_KIND_PORTABLE, relationsStorage);
+            context = serializationService.getPortableContext();
+        }
+
+        @Override
+        protected String getTypeMetadata(PortableId portableId) {
+            return portableId.toString();
+        }
+
+        @Override
+        protected ClassDefinition getSchema(PortableId portableId) {
+            ClassDefinition classDef = context.lookupClassDefinition(portableId);
+            if (classDef == null) {
+                throw QueryException.error("The given factoryId/classId/version combination not known to the member");
+            }
+            return classDef;
+        }
+
+        @Override
+        protected List<TypeField> resolveFields(ClassDefinition classDef) {
+            return classDef.getFieldNames().stream().map(name -> {
+                FieldDefinition field = classDef.getField(name);
+                if (field.getType().equals(FieldType.PORTABLE)) {
+                    throw QueryException.error("Column list is required to create nested fields");
+                }
+                return new TypeField(name, PORTABLE_TO_SQL.getOrDefault(field.getType()));
+            }).collect(toList());
+        }
+
+        @Override
+        protected PortableId getFieldSchemaId(ClassDefinition classDef, String fieldName, String fieldTypeName) {
+            return classDef.getField(fieldName).getPortableId();
+        }
+
+        @Override
+        protected PortableId getSchemaId(Map<String, String> options, Boolean isKey) {
+            return isKey != null ? portableId(options, isKey) : null;
+        }
+    }
+
+    private static class CompactEnricher extends FieldEnricher<String, Void> {
+        CompactEnricher(RelationsStorage relationsStorage) {
+            super(QueryDataType.OBJECT_TYPE_KIND_COMPACT, relationsStorage);
+        }
+
+        @Override
+        protected String getTypeMetadata(String compactTypeName) {
+            return compactTypeName;
+        }
+
+        @Override
+        protected Void getSchema(String schemaId) {
+            return null;
+        }
+
+        @Override
+        protected List<TypeField> resolveFields(Void schema) {
+            throw QueryException.error("Column list is required to create Compact-based types");
+        }
+
+        @Override
+        protected String getFieldSchemaId(Void schema, String fieldName, String fieldTypeName) {
+            return fieldTypeName + "CompactType";
+        }
+
+        @Override
+        protected String getSchemaId(Map<String, String> options, Boolean isKey) {
+            return options.get(isKey == null ? SqlConnector.OPTION_TYPE_COMPACT_TYPE_NAME : isKey
```

Comment:
nested ternary expressions tend to be hard to read. Does not sonar complain about it?

Reply:
How can I simplify a tristate boolean further? `condition == null ? ... : condition ? ... : ...` is, I think, the simplest version that can be written in Java.

Reply:
> isKey should be null if you pass type options and non-null for mapping options.

if so, [enum](https://thedailywtf.com/articles/What_Is_Truth_0x3f_) might be better, eg. `{KEY, VALUE, TYPE}`

Reply:
Actually, you are right. In the conventional semantics of "tristate boolean," the states are true, false and indeterminate. However, in my usage, there are three determinate states, so I actually abused `Boolean` having three states. I am switching to enum.